librad/src/profile.rs

```diff
@@ -0,0 +1,191 @@
+    /// and system paths, scoped to the identifier, are used. See
+    /// [`Paths::new`] for details.
+    ///
+    /// Otherwise and if the `RAD_HOME` environment variable is set, the value
+    /// is used to construct the profile with [`Profile::from_root`].
+    ///
+    /// Otherwise, the profile identifier is read from a well-known file and
+    /// system paths scoped to the identifier are used. If the file containing
+    /// the identifier  does not exist, it is created and a new identifier is
+    /// generated and written to the file.
+    ///
+    /// The profile identifier must not empty, must not contain path separators,
+    /// must not be a windows path prefix like `C:`, and must not be a
+    /// special component like `.` or `..`.
+    ///
+    /// On Linux, the path to the active profile is
+    /// `$XDG_CONFIG_HOME/radicle-link/active_profile`.
+    pub fn load() -> Result<Self, Error> {
+        if let Ok(id) = env::var("RAD_PROFILE") {
+            let paths = Paths::new(&id)?;
+            return Ok(Self { id, paths });
+        }
+
+        if let Ok(rad_home) = env::var("RAD_HOME") {
+            return Self::from_root(Path::new(&rad_home));
+        }
+
+        let id = load_profile_id(project_dirs()?.config_dir())?;
+
+        let paths = Paths::new(&id)?;
+
+        Ok(Self { id, paths })
+    }
+
+    /// Creates a profile where `root` is used as the root for all profile
+    /// directories. See [`Paths::from_root`] for details. The profile ID is
+    /// stored in `<root>/active_profile`.
+    pub fn from_root(root: &Path) -> Result<Self, Error> {
+        let id = load_profile_id(root)?;
+        let paths = Paths::from_root(root)?;
+
+        Ok(Self { id, paths })
+    }
+
+    /// Returns the profile identifier
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    /// Returns [`Paths`] for this profile.
+    pub fn paths(&self) -> &Paths {
+        &self.paths
+    }
+}
+
+/// Returns `true` if `id` is a valid profile ID.
+///
+/// A valid profile ID is not empty, does not contain path separators, is not
+/// a windows path prefix like `C:`, and is not a special component like `.` or
+/// `..`.
+fn is_valid_profile_id(id: &str) -> bool {
+    let mut components = Path::new(id).components();
+    if let Some(c) = components.next() {
+        match c {
+            path::Component::Normal(_) => {},
+            _ => return false,
+        }
+    } else {
+        return false;
+    };
+
+    if components.next().is_some() {
+        return false;
+    }
+
+    true
+}
+
+/// Read the profile ID from a file or create the file and write a newly
+/// generated profile ID to it.
+fn load_profile_id(config_dir: &Path) -> Result<String, Error> {
+    let active_profile_path = config_dir.join("active_profile");
+
+    match fs::read_to_string(&active_profile_path) {
+        Ok(content) => {
+            let id = content.lines().next().unwrap();
```

Comment:
Is this `unwrap` safe?

Reply:
Yes. Will add a comment. Or is the preferred way to put the explanation as the argument to `expect`?

Reply:
Right, I'd have expected the contents would need to be non-empty :) Sure the file could be there but what happens if it's empty :smile:

Reply:
I tested this and it threw an error for me. Maybe I removed the profile id in an incorrect way, but it doesn't seem safe to assume it being removed in a certain way. (What I did: opened the file with `vim`, used `dd` to remove the line, closed with `:x`)

Reply:
I think it makes sense to treat an empty (but present) `active_profile` file like there is no ID yet. Will update the PR.

Reply:
What kind of scenario would that be useful for?

Reply:
I think it’s just to avoid confusion on the users side. I can definitely foresee somebody deleting the content of the file and then wondering why profile is invalid. If you feel strongly otherwise, @FintanH, we can also leave as it is.

Reply:
The idea is to trigger a new Identity being created when no `active_profile` can be found. I think we should serve to both cases if the user removes the contents of the file or deletes the file completely.

Reply:
Ya, I would think this is an error variant. I wouldn't think that an empty profile is a valid one.

Reply:
There seems to be some disagreement here, @MeBrei, @FintanH. Either treat an empty file as an error or treat it as no active profile does exist.

After giving this some thought I believe it would be best to treat an empty file as an error. This makes the behavior more straight-forward and reduces the number of edge cases. If the user deletes the content of the file then we throw an error instead of working around this.

Reply:
Makes sense to me :ok_hand: